components/automate-ui/src/app/entities/userperms/userperms.effects.ts

```diff
@@ -17,27 +28,35 @@ import {
   GetUserParamPermsFailure
 } from './userperms.actions';
 import { UserPermsRequests } from './userperms.requests';
+import { getlastFetchTime } from './userperms.selectors';
 
 @Injectable()
 export class UserPermEffects {
   constructor(
     private actions$: Actions,
-    private requests: UserPermsRequests
+    private requests: UserPermsRequests,
+    private store: Store<NgrxStateAtom>
   ) { }
 
-  // include debounce here because each request of this type is a true duplicate
   @Effect()
   fetchAllPerms$ = this.actions$.pipe(
     ofType(UserPermsTypes.GET_ALL),
-    debounceTime(20),
-    mergeMap(() => this.requests.fetchAll().pipe(
-      map((resp: UserPermsResponsePayload) => new GetAllUserPermsSuccess(resp.endpoints)),
-      catchError((error: HttpErrorResponse) => of(new GetAllUserPermsFailure(error))))));
+    debounceTime(20), // each request of this type is a true duplicate so debounce them
```

Comment:
(3) Even if the dispatch fires a few times rapidly, this debounce ensures it only sends one occurrence through.

Reply:
`debounceTime` in effect causes a 20ms delay here -- since the first request `a` is kept in waiting for 20ms, before it's fired (to let any subsequent requests `b`, `c`, ... within those 20ms be ignored); and each subsequent requests *resets the 20ms*. This doesn't seem optimal for our use case here.

See https://rxjs-dev.firebaseapp.com/api/operators/debounceTime#description:

![img](https://rxjs-dev.firebaseapp.com/assets/images/marble-diagrams/debounceTime.png)

Do you know an alternative that fires `a` right away but ignores subsequent requests in a 20ms window?

Reply:
I originally put that in knowing it fires at the end of the interval, as you correctly state, as I felt it was short enough to be insignificant. However, since you raised the question I did come up with an improvement: https://rxjs-dev.firebaseapp.com/api/operators/throttleTime

Reply:
Cool, thanks. If we're using `throttleTime`, I suppose we could up the `20ms` _a lot_ and save some more requests...? 😃

Reply:
Already done, and turned it even further to an advantage using the throttle itself as the cache expiry for `IntrospectAll` itself.